src/utils/src/ExponentialBackoffRetryStrategy.c

```diff
@@ -0,0 +1,245 @@
+            CHK_ERR(FALSE, STATUS_EXPONENTIAL_BACKOFF_INVALID_STATE, "Exponential backoff is already terminated");
+            // No 'break' needed since CHK(FALSE, ...) will always jump to CleanUp
+        default:
+            DLOGD("Cannot execute exponentialBackoffBlockingWait. Unexpected state [%"PRIu64"]", pExponentialBackoffRetryStrategyState->status);
+            CHK_ERR(FALSE, STATUS_EXPONENTIAL_BACKOFF_INVALID_STATE, "Unexpected exponential backoff state");
+    }
+
+CleanUp:
+    LEAVES();
+    return retStatus;
+}
+
+STATUS exponentialBackoffRetryStrategyBlockingWait(PRetryStrategy pRetryStrategy) {
+    ENTERS();
+    PExponentialBackoffRetryStrategyState pRetryState = NULL;
+    PExponentialBackoffRetryStrategyConfig pRetryConfig = NULL;
+    UINT64 currentSystemTime = 0, currentRetryWaitTime = 0, jitter = 0;
+    STATUS retStatus = STATUS_SUCCESS;
+
+    CHK(pRetryStrategy != NULL, STATUS_NULL_ARG);
+
+    pRetryState = TO_EXPONENTIAL_BACKOFF_STATE(pRetryStrategy);
+    CHK_STATUS(validateAndUpdateExponentialBackoffStatus(pRetryState));
+    pRetryConfig = &(pRetryState->exponentialBackoffRetryStrategyConfig);
+
+    // If retries is exhausted, return error to the application
+    CHK(pRetryConfig->maxRetryCount == KVS_INFINITE_EXPONENTIAL_RETRIES ||
+         pRetryConfig->maxRetryCount > pRetryState->currentRetryCount, STATUS_EXPONENTIAL_BACKOFF_RETRIES_EXHAUSTED);
+
+    // check the last retry time. If the difference between last retry and current time is greater than
+    // minTimeToResetRetryState, then reset the retry state
+    //
+    // Proceed if this is not the first retry
+    currentSystemTime = GETTIME();
+    if (pRetryState->currentRetryCount != 0 &&
+        currentSystemTime > pRetryState->lastRetrySystemTime && // sanity check
+        currentSystemTime - pRetryState->lastRetrySystemTime > pRetryConfig->minTimeToResetRetryState) {
+        CHK_STATUS(resetExponentialBackoffRetryState(pRetryState));
+        pRetryState->status = BACKOFF_IN_PROGRESS;
+    }
+
+    // Bound the exponential curve to maxRetryWaitTime. Once we reach
+    // maxRetryWaitTime, then we always wait for maxRetryWaitTime time
+    // till the state is reset.
+    if (pRetryState->lastRetryWaitTime >= pRetryConfig->maxRetryWaitTime) {
+        currentRetryWaitTime = pRetryConfig->maxRetryWaitTime;
+    } else {
+        CHK_STATUS(calculateWaitTime(pRetryState, pRetryConfig, &currentRetryWaitTime));
+        if (currentRetryWaitTime > pRetryConfig->maxRetryWaitTime) {
+            currentRetryWaitTime = pRetryConfig->maxRetryWaitTime;
+        }
+    }
+
+    // Note: Do not move the call to getRandomJitter in calculateWaitTime.
+    // This is because we need randomization for wait time after we reach maxRetryWaitTime
+    jitter = getRandomJitter(pRetryState->exponentialBackoffRetryStrategyConfig.jitterFactor);
+    currentRetryWaitTime += jitter;
+    THREAD_SLEEP(currentRetryWaitTime);
```

Comment:
This is going to be scary. Would be nice to properly comment and document the top-level scenario when this is executed. For example, what scenario this will be used with the values. In case of create stream the wait will be on the thread of execution. The thread could come from the networking thread in C Producer. How long this will be blocked for? Another scenario is the token rotation - in this case the thread is the thread of the new session handlers PutStream API from c producer. WebRTC scenarios will also be two fold - the main execution thread and the network handler/listener thread.

Reply:
Yep. I agree. I did not have a good understanding of the threading model and it makes sense that we should not block the caller thread. I've incorporated this change in the latest commit. Please take a look.
Instead of callAfter, I've directly updated `pStateMachineImpl->context.time`.